Source/Canary/Mqtt5Canary/Mqtt5Canary.swift

```diff
@@ -0,0 +1,559 @@
+    )
+    print(
+      "Unsubscribe - Attempts: \(unsubAttempt), Success: \(unsubSucceed), Failed: \(unsubFailed), Success Rate: \(String(format: "%.2f", unsubscribeSuccessRate))%"
+    )
+  }
+
+}
+
+enum Mqtt5CanaryOperation: Int {
+  case NULL = 0
+  case START
+  case STOP
+  case DESTROY
+  case SUBSCRIBE
+  case UNSUBSCRIBE
+  case UNSUBSCRIBE_BAD
+  case PUBLISH_QOS0
+  case PUBLISH_QOS1
+  case PUBLISH_TO_SUBSCRIBED_TOPIC_QOS0
+  case PUBLISH_TO_SUBSCRIBED_TOPIC_QOS1
+  case PUBLISH_TO_SHARED_TOPIC_QOS0
+  case PUBLISH_TO_SHARED_TOPIC_QOS1
+  case OPERATION_COUNT
+}
+
+// This struct holds the canary options
+// Adjustification of "@uncheck Sendable": The memebers in this struct will not be modified once setup.
+struct Mqtt5CanaryTestOptions: @unchecked Sendable {
+  // Client options
+  let shared_topic: String
+  var elg: EventLoopGroup
+  var boostrap: ClientBootstrap
+  var tlsctx: TLSContext? = nil
+  // Test options
+  let tpsSleepTime: UInt64
+  var onWebsocketTransform: OnWebSocketHandshakeIntercept? = nil
+  var operationDistribution: [Mqtt5CanaryOperation] = []
+
+  init(testApp: Mqtt5Canary) throws {
+    shared_topic = "shared_topic_" + UUID().uuidString
+    self.tpsSleepTime = testApp.tps > 0 ? (UInt64(NANO_SECOND / testApp.tps)) : 0
+    // Initialize elg first
+    self.elg = try EventLoopGroup(threadCount: testApp.threads)
+    let resolver = try HostResolver(
+      eventLoopGroup: self.elg,
+      maxHosts: 8,
+      maxTTL: 30)
+    self.boostrap = try ClientBootstrap(
+      eventLoopGroup: self.elg,
+      hostResolver: resolver)
+
+    try self.setupClientOption(testApp: testApp)
+
+    // Setup operation distribution
+    Mqtt5CanaryOperationDistributionSetup(&self.operationDistribution)
+  }
+
+  mutating func setupClientOption(testApp: Mqtt5Canary) throws {
+    if (testApp.useTls && self.tlsctx == nil) {
+      guard let _cert = testApp.cert, let _key = testApp.key else {
+        throw CanaryTestError.InvalidArgument
+      }
+      let tlsOptions = try TLSContextOptions.makeMTLS(
+        certificatePath: _cert,
+        privateKeyPath: _key
+      )
+
+      if let _caPath = testApp.caPath {
+        try tlsOptions.overrideDefaultTrustStoreWithFile(caFile: _caPath)
+      }
+      self.tlsctx = try TLSContext(options: tlsOptions, mode: .client)
+    }
+
+    if (testApp.useWebsocket) {
+      self.onWebsocketTransform = { request, complete in
+        // Simply complete the request with success
+        complete(request, 0)
+      }
+    }
+  }
+}
+
+func Mqtt5CanaryOperationDistributionSetup(_ distributionDataSet: inout [Mqtt5CanaryOperation]) {
+  let operationDistribution = [
+    (Mqtt5CanaryOperation.STOP, 1),
+    (Mqtt5CanaryOperation.SUBSCRIBE, 200),
+    (Mqtt5CanaryOperation.UNSUBSCRIBE, 200),
+    (Mqtt5CanaryOperation.UNSUBSCRIBE_BAD, 100),
```

Comment:
Trivial: Since we may hit the bad unsub due to sub and unsub being equal, should we lower this further to prevent an excess of unsub bad? I don't think it really matters...